### PR TITLE
Decouple zod

### PR DESCRIPTION
**Pull request by diogob**
Decoupling from zod yields a clearer typing IMHO.
It will also allow anyone to build adapters for any validator as long as they abide by zod's `safeParseAsync` and issues interface.

The long term goal is to track Zod's interface keeping this library always as our de facto standard parser.
The additional burden on type maintenance should be minimal considering these interfaces have a small surface area and are quite fundamental to Zod's backwards compatibility.

## TODO

- [x] Double check if new types are tight enough.
- [x] Write JSDoc for exposed types.
- [x] Write note about adapters in README.

### CODE REVIEW

File: src/constructor.ts [hunk 1]
```diff
@@ -118,10 +111,31 @@ function makeDomainFunction<
...
       })
     } as DomainFunction<Awaited<Output>>
   }
 }
 
+// deno-lint-ignore ban-types
+const objectSchema: ParserSchema<{}> = {
+  // deno-lint-ignore require-await
+  safeParseAsync: async (data: unknown) => {
+    if (data == null || typeof data !== 'object') {
```
Comment (gustavoguichard) [inline src/constructor.ts:124]: `typeof [] === "object"` though..
This is the only safe way I know to assure that:
```ts
Object.prototype.toString.call({}) === "[object Object]"
Object.prototype.toString.call([]) === "[object Array]"
Object.prototype.toString.call(null) === "[object Null]"
```

Check out [this function](https://github.com/gustavoguichard/make-service/blob/main/src/internals.ts#L31)

**Review by gustavoguichard — COMMENTED** [review]
LGTM ❤️

File: src/constructor.ts (hunk 1 above)
Comment (diogob) [reply src/constructor.ts:124]: Yeah, I know the type was wrong (it is in the todo to review these types). But now I see the runtime check is wrong as well!

File: src/constructor.ts [hunk 2]
```diff
@@ -118,10 +111,31 @@ function makeDomainFunction<
...
         }
-        return handler(result.data, envResult.data)
+        return handler(result.data as I, envResult.data as E)
       })
     } as DomainFunction<Awaited<Output>>
   }
 }
 
+// deno-lint-ignore ban-types
+const objectSchema: ParserSchema<{}> = {
```
Comment (gustavoguichard) [inline src/constructor.ts:121]:
```suggestion
const objectSchema: ParserSchema<Record<PropertyKey, unknown>> = {
```

Comment (gustavoguichard) [reply src/constructor.ts:121]: In TS `{}` means anything but `null | undefined`

File: src/constructor.ts [hunk 3]
```diff
@@ -118,10 +111,31 @@ function makeDomainFunction<
...
+  // deno-lint-ignore require-await
```
Comment (gustavoguichard) [inline src/constructor.ts:122]: To remove this comment you can remove the `async` and return `Promise.resolve({ success: true, data })`

**Review by gustavoguichard — CHANGES_REQUESTED** [review]
@diogob now I realize that on `objectSchema` and `undefinedSchema`, in case of failure, it should return a `ParserIssue[]` instead of throwing, as they are using a safe method: `safeParseAsync`

And we should also document it ;)

**Comment by diogob** [comment]
> @diogob now I realize that on `objectSchema` and `undefinedSchema`, in case of failure, it should return a `ParserIssue[]` instead of throwing, as they are using a safe method: `safeParseAsync`
> 
> And we should also document it ;)

Great catch! I'm glad we never rush merges :P

**Comment by gustavoguichard** [comment]
I'm getting pretty excited about this PR tho

**Review by gustavoguichard — DISMISSED** [review]

**Pull request merged**